File: frontend/src/api/metadata.ts
/**
 * 元数据管理相关API
 */
import { apiClient } from './client'
import type { ApiResponse, PaginatedResponse } from './client'

// 元数据相关类型定义
export interface DataTheme {
  id: number
  name: string
  description?: string
  icon?: string
  color?: string
  sort_order: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface DataTable {
  id: number
  theme_id: number
  table_name: string
  table_alias: string
  table_description?: string
  is_active: boolean
  created_at: string
  updated_at: string
  theme?: DataTheme
}

export interface DataField {
  id: number
  table_id: number
  field_name: string
  field_alias: string
  field_description?: string
  data_type: string
  is_primary_key: boolean
  is_nullable: boolean
  default_value?: string
  is_active: boolean
  created_at: string
  updated_at: string
  table?: DataTable
}

export interface GlossaryTerm {
  id: number
  term_name: string
  term_description: string
  category?: string
  is_active: boolean
  created_at: string
  updated_at: string
}

// API请求参数类型
export interface ThemeListParams {
  page?: number
  size?: number
  keyword?: string
  is_active?: boolean
}

export interface TableListParams {
  page?: number
  size?: number
  theme_id?: number
  keyword?: string
  is_active?: boolean
}

export interface FieldListParams {
  page?: number
  size?: number
  table_id?: number
  keyword?: string
  is_active?: boolean
}

export interface GlossaryListParams {
  page?: number
  size?: number
  keyword?: string
  category?: string
  is_active?: boolean
}

/**
 * 元数据API类
 */
export class MetadataApi {
  // 数据主题相关
  async getThemes(params?: ThemeListParams): Promise<ApiResponse<DataTheme[]>> {
    const response = await apiClient.get('/metadata/themes', { params })
    // 从分页响应中提取items
    return {
      ...response,
      data: response.data.items || []
    }
  }

  async getTheme(id: number): Promise<ApiResponse<DataTheme>> {
    return apiClient.get(`/metadata/themes/${id}`)
  }

  async createTheme(data: Partial<DataTheme>): Promise<ApiResponse<DataTheme>> {
    return apiClient.post('/metadata/themes', data)
  }

  async updateTheme(id: number, data: Partial<DataTheme>): Promise<ApiResponse<DataTheme>> {
    return apiClient.put(`/metadata/themes/${id}`, data)
  }

  async deleteTheme(id: number): Promise<ApiResponse<void>> {
    return apiClient.delete(`/metadata/themes/${id}`)
  }

  // 数据表相关
  async getTables(params?: TableListParams): Promise<ApiResponse<PaginatedResponse<DataTable>>> {
    return apiClient.get('/metadata/tables', { params })
  }

  async getTable(id: number): Promise<ApiResponse<DataTable>> {
    return apiClient.get(`/metadata/tables/${id}`)
  }

  async getTablesByTheme(themeId: number): Promise<ApiResponse<DataTable[]>> {
    const response = await apiClient.get('/metadata/tables', { 
      params: { theme_id: themeId, size: 100 } 
    })
    // 从分页响应中提取items
    return {
      ...response,
      data: response.data.items || []
    }
  }

  async createTable(data: Partial<DataTable>): Promise<ApiResponse<DataTable>> {
    return apiClient.post('/metadata/tables', data)
  }

  async updateTable(id: number, data: Partial<DataTable>): Promise<ApiResponse<DataTable>> {
    return apiClient.put(`/metadata/tables/${id}`, data)
  }

  async deleteTable(id: number): Promise<ApiResponse<void>> {
    return apiClient.delete(`/metadata/tables/${id}`)
  }

  // 数据字段相关
  async getFields(params?: FieldListParams): Promise<ApiResponse<PaginatedResponse<DataField>>> {
    return apiClient.get('/metadata/fields', { params })
  }

  async getField(id: number): Promise<ApiResponse<DataField>> {
    return apiClient.get(`/metadata/fields/${id}`)
  }

  async createField(data: Partial<DataField>): Promise<ApiResponse<DataField>> {
    return apiClient.post('/metadata/fields', data)
  }

  async updateField(id: number, data: Partial<DataField>): Promise<ApiResponse<DataField>> {
    return apiClient.put(`/metadata/fields/${id}`, data)
  }

  async deleteField(id: number): Promise<ApiResponse<void>> {
    return apiClient.delete(`/metadata/fields/${id}`)
  }

  // 术语词汇相关
  async getGlossaryTerms(params?: GlossaryListParams): Promise<ApiResponse<PaginatedResponse<GlossaryTerm>>> {
    return apiClient.get('/metadata/glossary', { params })
  }

  async getGlossaryTerm(id: number): Promise<ApiResponse<GlossaryTerm>> {
    return apiClient.get(`/metadata/glossary/${id}`)
  }

  async createGlossaryTerm(data: Partial<GlossaryTerm>): Promise<ApiResponse<GlossaryTerm>> {
    return apiClient.post('/metadata/glossary', data)
  }

  async updateGlossaryTerm(id: number, data: Partial<GlossaryTerm>): Promise<ApiResponse<GlossaryTerm>> {
    return apiClient.put(`/metadata/glossary/${id}`, data)
  }

  async deleteGlossaryTerm(id: number): Promise<ApiResponse<void>> {
    return apiClient.delete(`/metadata/glossary/${id}`)
  }

  // 元数据统计
  async getMetadataStats(): Promise<ApiResponse<{
    theme_count: number
    table_count: number
    field_count: number
    glossary_count: number
  }>> {
    return apiClient.get('/metadata/stats')
  }

  // 元数据搜索
  async searchMetadata(keyword: string): Promise<ApiResponse<{
    themes: DataTheme[]
    tables: DataTable[]
    fields: DataField[]
    glossary: GlossaryTerm[]
  }>> {
    return apiClient.get('/metadata/search', { params: { keyword } })
  }
}

// 导出API实例
export const metadataApi = new MetadataApi()

// 默认导出
export default metadataApi

File: backend/api/v1/metadata.py
"""
元数据管理 API 路由
"""
from typing import List
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from utils.database import get_db
from utils.exceptions import ValidationException
from schemas.base import DataResponse, PaginatedResponse
from schemas.metadata_schemas import (
    DataThemeCreate, DataThemeUpdate, DataThemeResponse, DataThemeListParams,
    TableCreate, TableUpdate, TableResponse, TableListParams,
    FieldCreate, FieldUpdate, FieldResponse, FieldListParams,
    GlossaryCreate, GlossaryUpdate, GlossaryResponse, GlossaryListParams,
    RelationCreate, RelationUpdate, RelationResponse
)
from services.metadata_service import get_metadata_service

router = APIRouter()


# ========== 数据主题管理 ==========
@router.post("/themes", response_model=DataResponse[DataThemeResponse], summary="创建数据主题")
async def create_theme(
    theme_data: DataThemeCreate,
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """创建数据主题"""
    service = get_metadata_service(db)
    theme = await service.create_theme(theme_data, current_user_id)
    return DataResponse(data=theme)


@router.get("/themes/{theme_id}", response_model=DataResponse[DataThemeResponse], summary="获取数据主题详情")
async def get_theme(
    theme_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取数据主题详情"""
    service = get_metadata_service(db)
    theme = await service.get_theme(theme_id)
    return DataResponse(data=theme)


@router.put("/themes/{theme_id}", response_model=DataResponse[DataThemeResponse], summary="更新数据主题")
async def update_theme(
    theme_id: int,
    theme_data: DataThemeUpdate,
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """更新数据主题"""
    service = get_metadata_service(db)
    theme = await service.update_theme(theme_id, theme_data, current_user_id)
    return DataResponse(data=theme)


@router.delete("/themes/{theme_id}", response_model=DataResponse[bool], summary="删除数据主题")
async def delete_theme(
    theme_id: int,
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """删除数据主题"""
    service = get_metadata_service(db)
    result = await service.delete_theme(theme_id, current_user_id)
    return DataResponse(data=result)


@router.get("/themes", response_model=PaginatedResponse[DataThemeResponse], summary="获取数据主题列表")
async def list_themes(
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    keyword: str = Query(None, description="搜索关键词"),
    is_public: bool = Query(None, description="是否公共主题"),
    sort_by: str = Query("created_at", description="排序字段"),
    sort_order: str = Query("desc", regex="^(asc|desc)$", description="排序方向"),
    db: AsyncSession = Depends(get_db)
):
    """获取数据主题列表"""
    params = {
        'page': page,
        'size': size,
        'keyword': keyword,
        'is_public': is_public,
        'sort_by': sort_by,
        'sort_order': sort_order
    }
    
    service = get_metadata_service(db)
    result = await service.list_themes(params)
    return PaginatedResponse(data=result)


# ========== 表元数据管理 ==========
@router.post("/tables", response_model=DataResponse[TableResponse], summary="创建表元数据")
async def create_table(
    table_data: TableCreate,
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """创建表元数据"""
    service = get_metadata_service(db)
    table = await service.create_table(table_data, current_user_id)
    return DataResponse(data=table)


@router.get("/tables/{table_id}", response_model=DataResponse[TableResponse], summary="获取表元数据详情")
async def get_table(
    table_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取表元数据详情"""
    service = get_metadata_service(db)
    table = await service.get_table(table_id)
    return DataResponse(data=table)


@router.put("/tables/{table_id}", response_model=DataResponse[TableResponse], summary="更新表元数据")
async def update_table(
    table_id: int,
    table_data: TableUpdate,
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """更新表元数据"""
    service = get_metadata_service(db)
    table = await service.update_table(table_id, table_data, current_user_id)
    return DataResponse(data=table)


@router.get("/tables", response_model=PaginatedResponse[TableResponse], summary="获取表元数据列表")
async def list_tables(
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    keyword: str = Query(None, description="搜索关键词"),
    data_source: str = Query(None, description="数据源"),
    theme_id: int = Query(None, description="主题ID"),
    is_active: bool = Query(None, description="是否启用"),
    sort_by: str = Query("created_at", description="排序字段"),
    sort_order: str = Query("desc", regex="^(asc|desc)$", description="排序方向"),
    db: AsyncSession = Depends(get_db)
):
    """获取表元数据列表"""
    params = {
        'page': page,
        'size': size,
        'keyword': keyword,
        'data_source': data_source,
        'theme_id': theme_id,
        'is_active': is_active,
        'sort_by': sort_by,
        'sort_order': sort_order
    }
    
    # TODO: 实现 list_tables 方法
    # service = get_metadata_service(db)
    # result = await service.list_tables(params)
    # return PaginatedResponse(data=result)
    
    # 临时返回空结果
    from schemas.base import PaginatedData
    empty_result = PaginatedData.create([], 0, page, size)
    return PaginatedResponse(data=empty_result)


# ========== 字段元数据管理 ==========
@router.post("/fields", response_model=DataResponse[FieldResponse], summary="创建字段元数据")
async def create_field(
    field_data: FieldCreate,
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """创建字段元数据"""
    service = get_metadata_service(db)
    field = await service.create_field(field_data, current_user_id)
    return DataResponse(data=field)


@router.get("/fields/{field_id}", response_model=DataResponse[FieldResponse], summary="获取字段元数据详情")
async def get_field(
    field_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取字段元数据详情"""
    # TODO: 实现 get_field 方法
    from utils.exceptions import ResourceNotFoundException
    raise ResourceNotFoundException("字段详情接口待实现")


@router.get("/fields", response_model=PaginatedResponse[FieldResponse], summary="获取字段元数据列表")
async def list_fields(
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    table_id: int = Query(None, description="表ID"),
    keyword: str = Query(None, description="搜索关键词"),
    field_type: str = Query(None, description="字段类型"),
    is_active: bool = Query(None, description="是否启用"),
    db: AsyncSession = Depends(get_db)
):
    """获取字段元数据列表"""
    # TODO: 实现 list_fields 方法
    from schemas.base import PaginatedData
    empty_result = PaginatedData.create([], 0, page, size)
    return PaginatedResponse(data=empty_result)


# ========== 术语管理 ==========
@router.post("/glossary", response_model=DataResponse[GlossaryResponse], summary="创建术语")
async def create_glossary(
    glossary_data: GlossaryCreate,
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """创建术语"""
    service = get_metadata_service(db)
    glossary = await service.create_glossary(glossary_data, current_user_id)
    return DataResponse(data=glossary)


@router.get("/glossary/{glossary_id}", response_model=DataResponse[GlossaryResponse], summary="获取术语详情")
async def get_glossary(
    glossary_id: int,
    db: AsyncSession = Depends(get_db)
):
    """获取术语详情"""
    # TODO: 实现 get_glossary 方法
    from utils.exceptions import ResourceNotFoundException
    raise ResourceNotFoundException("术语详情接口待实现")


@router.get("/glossary", response_model=PaginatedResponse[GlossaryResponse], summary="获取术语列表")
async def list_glossary(
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    keyword: str = Query(None, description="搜索关键词"),
    term_type: str = Query(None, description="术语类型"),
    category: str = Query(None, description="术语分类"),
    db: AsyncSession = Depends(get_db)
):
    """获取术语列表"""
    # TODO: 实现 list_glossary 方法
    from schemas.base import PaginatedData
    empty_result = PaginatedData.create([], 0, page, size)
    return PaginatedResponse(data=empty_result)


# ========== 数据同步和导入 ==========
@router.post("/sync/tables", response_model=DataResponse[dict], summary="同步表结构")
async def sync_tables(
    data_source: str = Query(..., description="数据源名称"),
    schema_name: str = Query(None, description="模式名称"),
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """从数据源同步表结构"""
    # TODO: 实现表结构同步功能
    return DataResponse(data={"message": "表结构同步功能待实现", "data_source": data_source})


@router.post("/import/metadata", response_model=DataResponse[dict], summary="导入元数据")
async def import_metadata(
    # file: UploadFile = File(..., description="元数据文件"),
    db: AsyncSession = Depends(get_db),
    current_user_id: int = 1  # TODO: 从认证中间件获取
):
    """从文件导入元数据"""
    # TODO: 实现元数据导入功能
    return DataResponse(data={"message": "元数据导入功能待实现"})


@router.get("/export/metadata", summary="导出元数据")
async def export_metadata(
    format: str = Query("json", regex="^(json|excel|csv)$", description="导出格式"),
    theme_ids: str = Query(None, description="主题ID列表，逗号分隔"),
    db: AsyncSession = Depends(get_db)
):
    """导出元数据"""
    # TODO: 实现元数据导出功能
    return {"message": "元数据导出功能待实现", "format": format}